autorest/codegen/__init__.py

```diff
@@ -135,12 +91,138 @@ def _build_package_dependency() -> Dict[str, str]:
+                "or 'embedded'"
+            )
+
+        if options["models_mode"] not in ["msrest", "none"]:
+            raise ValueError(
+                "--models-mode can only be 'msrest' or 'none'. "
+                "Pass in 'msrest' if you want msrest models, or "
+                "'none' if you don't want any."
+            )
+
+        if not options["show_operations"] and options["builders_visibility"] == "embedded":
+            raise ValueError(
+                "Can not embed builders without operations. "
+                "Either set --show-operations to True, or change the value of --builders-visibility "
+                "to 'public' or 'hidden'."
+            )
+
+        if not options["show_operations"] and options["add_python3_operation_files"]:
+            raise ValueError(
+                "Can not add typed sync operation files if you are not showing operations. "
+                "If you want typed synced operation files, you have to add flag "
+                "--show-operations"
+            )
+
+        if options["basic_setup_py"] and not options["package_version"]:
+            raise ValueError("--basic-setup-py must be used with --package-version")
+
+        if options["package_mode"] and not options["package_version"]:
+            raise ValueError("--package-mode must be used with --package-version")
+
+        if not options["show_operations"] and options["combine_operation_files"]:
+            raise ValueError(
+                "Can not combine operation files if you are not showing operations. "
+                "If you want operation files, pass in flag --show-operations"
+            )
+
+        if options["package_mode"]:
+            if options["package_mode"] not in ("mgmtplane", "dataplane") and not Path(options["package_mode"]).exists():
+                raise ValueError(
+                    "--package-mode can only be 'mgmtplane' or 'dataplane' or directory which contains template files"
+                )
+
+        if options["reformat_next_link"] and options["version_tolerant"]:
+            raise ValueError(
+                "--reformat-next-link can not be true for version tolerant generations. "
+                "Please remove --reformat-next-link from your call for version tolerant generations."
+            )
+
+        credential_scopes = self._get_credential_scopes()
+        if credential_scopes and not options["credential"]:
+            raise ValueError("--credential-scopes must be used with the --add-credential flag")
+
+        # check to see if user just passes in --credential-scopes with no value
+        if not credential_scopes and self._autorestapi.get_boolean_value("credential-scopes", False):
+            raise ValueError(
+                "--credential-scopes takes a list of scopes in comma separated format. "
+                "For example: --credential-scopes=https://cognitiveservices.azure.com/.default"
```

Comment:
I think it is better to move the check forward

Reply:
in the future, one thing to consider is making sure PRs are easier to read so errors / everything can be easily gone over, so I find it's best to not make huge diffs unless necessary. For example, here you could make the original function stay where it is and just pass it credential scopes so the diff is easier, and you're not adding more methods to the `CodeGenerator` class than necessary

This time i went ahead and checked and I think you're just adding credential scopes checks here

Reply:
https://github.com/Azure/autorest.python/pull/1161#issuecomment-1067543568